### ISSUE TITLE
Parallel-Chudnovsky is slow.

### ISSUE THREAD

**Issue by rsm-ltd** [davidcarver/Parallel-Chudnovsky-PI#2]
Why is parallel Chudnovsky so slow?

This is not a bug report as such but rather an observation.

The code on which Parallel-Chudnovsky-PI is based was originally written in 2002 by Hanhong Xue https://gmplib.org/download/misc/gmp-chudnovsky.c. It is a single threaded program.

What I am seeing is that Parallel-Chudnovsky  is only marginally faster than the original sequential version.

Here are my results when running on a dual core x86_64 box:

    $ time ./gmp-chudnovsky 50000000 0 
    #terms=3525683, depth=23
    sieve   time =  1.240
    ..................................................
    
    bs      time = 83.888
       gcd  time =  0.000
    div     time =  9.916
    sqrt    time =  5.452
    mul     time =  3.768
    total   time = 104.324
       P size=72801344 digits (1.456027)
       Q size=72801337 digits (1.456027)
    
    real    1m45.287s
    user    1m44.324s
    sys     0m0.888s

    $ time ./raspberry-pi2 50000000 0 1 
    #terms=3525683, depth=23, threads=1 cores=2
    bs1      cputime = 116.75  wallclock = 117.42   factor =    1.0
    bs2      cputime =   0.00  wallclock =   0.00   factor =    0.0
    div/sqrt cputime =  19.10  wallclock =  19.40   factor =    1.0
    mul      cputime =   3.78  wallclock =   3.83   factor =    1.0
    total    cputime = 139.73  wallclock = 140.77   factor =    1.0
       P size=121205329 digits (2.424107)
       Q size=121205323 digits (2.424106)

    real    2m20.774s
    user    2m19.732s
    sys     0m0.920s

    $ time ./raspberry-pi2 50000000 0 2 
    #terms=3525683, depth=23, threads=2 cores=2
    bs1      cputime = 106.52  wallclock =  62.41   factor =    1.7
    bs2      cputime =  15.72  wallclock =  16.19   factor =    1.0
    div/sqrt cputime =  19.53  wallclock =  11.74   factor =    1.7
    mul      cputime =   3.77  wallclock =   3.84   factor =    1.0
    total    cputime = 145.63  wallclock =  94.28   factor =    1.5
       P size=121205329 digits (2.424107)
       Q size=121205323 digits (2.424106)

    real    1m34.292s
    user    2m25.628s
    sys     0m1.960s
 
In summary:

Original sequential: 105 seconds - call that 100%
With OMP, 1 core:    140 seconds - 33% slow down!
With OMP, 2 core:     94 seconds - 11% speed up

In short we get 11% speed improvement at the cost of double the power consumption!

I would like to try this comparison on a four core machine but the only one I have available is the Pi 2 and gmp-chudnovsky fails there with a floating point exception.

It's interesting that calculating PI seems impervious to parallel execution. Since 2009 all the world records for calculating the digits of PI have been held by fairly normal home PC's, not the mega parallel supercomputers we have now a days.

**Comment by rsm-ltd** [comment]
I managed to get some comparative results for the original sequential and new parallel code running on the PI 2 and using 1 to 4 cores. 

The summary for 5 million digit calculations is:

    Original GMP    : 67 seconds   100%
    Parallel, 1 core: 87 seconds    30% Slower
    Parallel, 2 core: 55 seconds    20% Faster
    Parallel, 3 core: 49 seconds    26% Faster
    Parallel, 4 core: 41 seconds    40% Faster.

So we could say that two cores gets us a 20% speed up at the cost of twice the fuel consumption!

By the time we get to 4 cores we are only able to make use of a few percent of our extra cores power


    $ time ./raspberry-pi2 5000000 0 1
    #terms=352568, depth=20, threads=1 cores=4
    bs1      cputime =  69.62  wallclock =  70.07   factor =    1.0
    bs2      cputime =   0.00  wallclock =   0.00   factor =    0.0
    div/sqrt cputime =  14.49  wallclock =  14.72   factor =    1.0
    mul      cputime =   2.61  wallclock =   2.63   factor =    1.0
    total    cputime =  86.79  wallclock =  87.49   factor =    1.0
       P size=11062827 digits (2.212565)
       Q size=11062821 digits (2.212564)

    real    1m27.504s
    user    1m26.790s
    sys     0m0.700s
    
    $ time ./raspberry-pi2 5000000 0 2
    #terms=352568, depth=20, threads=2 cores=4
    bs1      cputime =  63.26  wallclock =  32.83   factor =    1.9
    bs2      cputime =  10.17  wallclock =  10.40   factor =    1.0
    div/sqrt cputime =  14.94  wallclock =   8.76   factor =    1.7
    mul      cputime =   2.59  wallclock =   2.63   factor =    1.0
    total    cputime =  91.03  wallclock =  54.69   factor =    1.7
       P size=11062827 digits (2.212565)
       Q size=11062821 digits (2.212564)

    real    0m54.708s
    user    1m31.030s
    sys     0m0.900s
    
    $ time ./raspberry-pi2 5000000 0 3
    #terms=352568, depth=20, threads=3 cores=4
    bs1      cputime =  57.93  wallclock =  20.98   factor =    2.8
    bs2      cputime =  16.57  wallclock =  16.78   factor =    1.0
    div/sqrt cputime =  14.94  wallclock =   8.74   factor =    1.7
    mul      cputime =   2.61  wallclock =   2.68   factor =    1.0
    total    cputime =  92.11  wallclock =  49.24   factor =    1.9
       P size=11062827 digits (2.212565)
       Q size=11062821 digits (2.212564)

    real    0m49.256s
    user    1m32.110s
    sys     0m0.990s

    $ time ./raspberry-pi2 5000000 0 4
    #terms=352568, depth=20, threads=4 cores=4
    bs1      cputime =  55.38  wallclock =  14.64   factor =    3.8
    bs2      cputime =  20.33  wallclock =  15.60   factor =    1.3
    div/sqrt cputime =  14.71  wallclock =   8.63   factor =    1.7
    mul      cputime =   2.56  wallclock =   2.63   factor =    1.0
    total    cputime =  93.02  wallclock =  41.56   factor =    2.2
       P size=11062827 digits (2.212565)
       Q size=11062821 digits (2.212564)
    
    real    0m41.577s
    user    1m33.030s
    sys     0m0.960s
    pi@raspberrypi ~/Parallel-Chudnovsky-PI $
